components/brave_wallet/browser/solana_instruction_data_decoder.cc

```diff
@@ -0,0 +1,882 @@
+  if (ret && *ret <= kAuthorityTypeMax)
+    return base::NumberToString(*ret);
+  return absl::nullopt;
+}
+
+absl::optional<uint32_t> DecodeUint32(const std::vector<uint8_t>& input,
+                                      size_t& offset) {
+  if (offset >= input.size() || input.size() - offset < 4) {
+    return absl::nullopt;
+  }
+
+  // Read bytes in little endian order.
+  uint32_t uint32_le = 0;
+  for (size_t i = offset; i < offset + 4; i++) {
+    uint32_le <<= 8;
+    uint32_le |= input[i];
+  }
+
+  offset += 4;
+
+#if defined(ARCH_CPU_LITTE_ENDIAN)
+  return uint32_le;
+#else
+  return base::ByteSwap(uint32_le);
+#endif
+}
+
+absl::optional<std::string> DecodeUint32String(
+    const std::vector<uint8_t>& input,
+    size_t& offset) {
+  auto ret = DecodeUint32(input, offset);
+  if (!ret)
+    return absl::nullopt;
+  return base::NumberToString(*ret);
+}
+
+absl::optional<uint64_t> DecodeUint64(const std::vector<uint8_t>& input,
+                                      size_t& offset) {
+  if (offset >= input.size() || input.size() - offset < 8) {
+    return absl::nullopt;
+  }
+
+  // Read bytes in little endian order.
+  uint64_t uint64_le = 0;
+  for (size_t i = offset; i < offset + 8; i++) {
+    uint64_le <<= 8;
+    uint64_le |= input[i];
+  }
+
+  offset += 8;
+
+#if defined(ARCH_CPU_LITTE_ENDIAN)
+  return uint64_le;
+#else
+  return base::ByteSwap(uint64_le);
+#endif
+}
+
+absl::optional<std::string> DecodeUint64String(
+    const std::vector<uint8_t>& input,
+    size_t& offset) {
+  auto ret = DecodeUint64(input, offset);
+  if (!ret)
+    return absl::nullopt;
+  return base::NumberToString(*ret);
+}
+
+absl::optional<std::string> DecodePublicKey(const std::vector<uint8_t>& input,
+                                            size_t& offset) {
+  if (offset >= input.size() || input.size() - offset < kSolanaPubkeySize)
+    return absl::nullopt;
+
+  offset += kSolanaPubkeySize;
+  return Base58Encode(std::vector<uint8_t>(
+      input.begin() + offset - kSolanaPubkeySize, input.begin() + offset));
+}
+
+absl::optional<std::string> DecodeOptionalPublicKey(
+    const std::vector<uint8_t>& input,
+    size_t& offset) {
+  if (offset == input.size()) {
+    return absl::nullopt;
+  }
+
+  // First byte is 0 or 1 to indicate if public key is passed.
+  // And the rest bytes are the actual public key.
+  if (input[offset] == 0) {
+    offset++;
+    return "";  // No public key is passed.
+  } else if (input[offset] == 1) {
+    offset++;
+    return DecodePublicKey(input, offset);
+  } else {
+    return absl::nullopt;
+  }
+}
+
+absl::optional<std::string> DecodeString(const std::vector<uint8_t>& input,
+                                         size_t& offset) {
+  // Decode u32 string len and u32 string len padding.
+  auto len = DecodeUint32(input, offset);
+  if (!len)
+    return absl::nullopt;
+  auto len_padding = DecodeUint32(input, offset);
+  if (!len_padding || *len_padding != 0) {
```

Comment:
can we add a comment saying the reason we need to check "padding" must equals 0? Because when "padding" has value, it means its length exceeds maximum 32 bytes, ex. `0x0100000000`( This will be `[0, 0, 0, 0, 1, 0, 0, 0]`) 
`len` and `len padding` here is kind of confusing here because `bincode::serialize` uses two u32 together to represent the length of the string, maybe we need a better naming?

Reply:
https://github.com/brave/brave-core/pull/15142/commits/b74bf46e583c5f4d6a757a90fda2fec21b1709f9